### ISSUE TITLE
Which are the best way to create a debug class

### ISSUE THREAD

**Issue by rafaelfigueiredo-l** [DevinVinson/WordPress-Plugin-Boilerplate#493]
Hi All,

First of all I need to thank you guys for this plugin. I'm study a lot and increase my programming using thig plugin. Thanks a lot!

Second I have an issue. I'm feeling the need to have a debug class inside my plugin so I created a class-debug.php inside the folder includes, and then create an instance like the boilerplate is using to create the loader class.

I need to use this debug class inside all my files (admin and public) and here is the question. Which are the bestest way to use this class inside all other files?

I created a static method, but I think it's not the bestest way. Should I recover an instance of this debug class inside all my other files and then use it?

Thank you all guys!

**Comment by sailorsamoor** [comment]
Hello.
I don't know what is the best way to create a debug class, but I create a different class and then create an instance. 

Best regards.

**Comment by dingo-d** [comment]
What do you need the debug class for? Debugging is only done when developing, and the usage of `error_log`, `print_r` or `var_dump` in a production is discouraged.

Also, _the best_ is already a superlative, so there is no _the bestest_ 😄

**Comment by sailorsamoor** [comment]
Not the fact. I use different debug class. There all of the debug functions are covered for best convenience.

**Comment by dingo-d** [comment]
Not saying you cannot use it, but I don't see the point in it...

**Comment by rafael-figueiredogluz** [comment]
@dingo-d I'm using elementor to create the entire page and I think with this approach all default theme hooks are gone. So, I created this new debug class to print and log my errors. But Before my plugin goes to production I will desactivate the print errors. Should I desactivate the logs too?

**Comment by rafael-figueiredogluz** [comment]
@sailorsamoor @dingo-d 

Thanks for your answers. I followed with an instance (singleton) and inside my plugin I'm using static methods.

**Comment by dingo-d** [comment]
Depends where you are logging them and when. If you are logging them in the `debug.log`, and only if the `WP_DEBUG` is set to true then go for it :) 

Although I'd avoid a singleton and static methods. They are a PITA to unit/integration test. Use dependency injection instead :)

**Comment by rafael-figueiredogluz** [comment]
@dingo-d Alright! I checked my code and I'm using Wp_Error class as a dependency injection inside my class. So, the log only works with WP_DEBUG active.

I googled it more about dependency injection and I think it's the right thing to change to this approach. 
I'm still not using tests, but I'm looking forward to learn. 

Altough I have a question. If I'm using this debug class to be used inside my entire plugin would it not be better to ensure that there are only one instance of this class? Because of this I used singleton and static methods, because my methods refers to the class and not to the instance.

Thanks for your answers! 👍
